### PR TITLE
refactor subnet gateway

### LINKED ISSUE #2747: Subnet gateway refactor

### PR DESCRIPTION
**Pull request by hongzhen-ma**
- [ ] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
- Refactor

### Which issue(s) this PR fixes:
Fixes #2747 

### <samp>🤖 Generated by Copilot at 6b8d7a9</samp>

Add subnet validation for logicalGateway and u2oInterconnection fields in `pkg/util/validator.go`. Prevent network conflicts by rejecting invalid subnet specs.

### <samp>🤖 Generated by Copilot at 6b8d7a9</samp>

> _`subnet` resource_
> _validates its fields with care_
> _no conflict in spring_

### <samp>🤖 Generated by Copilot at 6b8d7a9</samp>

* Add a validation check for subnet resource to prevent logicalGateway and u2oInterconnection conflicts ([link](https://github.com/kubeovn/kube-ovn/pull/2872/files?diff=unified&w=0#diff-7f0dbc597e1344dab5f2b581992a7ad0423ff2087593902fe39429d0a980782bR108-R112))

### CODE REVIEW

File: pkg/controller/subnet.go [hunk 1]
```diff
@@ -694,6 +596,125 @@ func (c *Controller) handleAddOrUpdateSubnet(key string) error {
...
+		return err
+	}
+	klog.V(4).Infof("handle add or update subnet %s", cachedSubnet.Name)
+
+	subnet := cachedSubnet.DeepCopy()
+	if err = formatSubnet(subnet, c); err != nil {
+		return err
+	}
+
+	subnet, err = c.subnetsLister.Get(key)
```
Comment (oilbeater) [inline pkg/controller/subnet.go:620]: Why get subnet again here?

File: pkg/controller/subnet.go [hunk 2]
```diff
@@ -1316,23 +1337,273 @@ func (c *Controller) reconcileVpcDelNormalStaticRoute(vpcName string) error {
...
+		klog.Infof("delete old centralized policy route for subnet %s", subnet.Name)
+		if err := c.deletePolicyRouteForCentralizedSubnet(subnet); err != nil {
+			klog.Errorf("failed to delete policy route for subnet %s, %v", subnet.Name, err)
+			return err
+		}
+		subnet.Spec.GatewayNode = ""
+		subnet.Status.ActivateGateway = ""
+		c.recorder.Eventf(subnet, v1.EventTypeNormal, "ChangeToDistributedGw", "")
+
+		if _, err := c.config.KubeOvnClient.KubeovnV1().Subnets().Update(context.Background(), subnet, metav1.UpdateOptions{}); err != nil {
```
Comment (oilbeater) [inline pkg/controller/subnet.go:1351]: should also update status

File: pkg/controller/subnet.go [hunk 3]
```diff
@@ -1316,23 +1337,273 @@ func (c *Controller) reconcileVpcDelNormalStaticRoute(vpcName string) error {
...
+	if err != nil {
+		klog.Errorf("failed to list nodes: %v", err)
+		return err
+	}
+	for _, node := range nodes {
+		if err = c.createPortGroupForDistributedSubnet(node, subnet); err != nil {
+			klog.Errorf("failed to create port group %v", err)
+			return err
+		}
+		if node.Annotations[util.AllocatedAnnotation] != "true" {
```
Comment (oilbeater) [inline pkg/controller/subnet.go:1367]: 如果node在这个时刻没有这个 annotation 会不会之后永远同步不到

File: pkg/controller/subnet.go [hunk 4]
```diff
@@ -1316,23 +1337,273 @@ func (c *Controller) reconcileVpcDelNormalStaticRoute(vpcName string) error {
...
+					return err
+				}
+			} else {
+				podName := c.getNameByPod(pod)
+				portName := ovs.PodNameToPortName(podName, pod.Namespace, podNet.ProviderName)
+				podPorts = append(podPorts, portName)
+			}
+		}
+
+		if pod.Annotations[util.NorthGatewayAnnotation] != "" {
```
Comment (oilbeater) [inline pkg/controller/subnet.go:1425]: northgatewayAnnotation和上面的判断合并

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/controller/subnet.go (hunk 4 above)
Comment (hongzhen-ma) [reply pkg/controller/subnet.go:1425]: 上边是在 pod 需要申请地址的 podNets 循环中，确认 pod 指定 NorthGateway，添加静态路由处理的。

循环外的判断，是确认已经处理完了指定 NorthGateway annotation的情况，不需要再继续往下了。在上边的 循环中 不好判断什么时候应该返回。

File: pkg/controller/subnet.go (hunk 2 above)
Comment (hongzhen-ma) [reply pkg/controller/subnet.go:1349]: 再次验证确认了下，update 确实只能更新 spec 字段。
已经修改，补充了 status 的更新

**Review by oilbeater — APPROVED** [review]

**Pull request merged**